File: app/cli/zig/src/eip191.zig
//! EIP-191 `personal_sign` (PROTOCOL.md §4).
//!
//! digest = keccak256(0x19 ‖ "Ethereum Signed Message:\n" ‖ decimal(byte_len) ‖ utf8(msg))
//!
//! The length is the UTF-8 **byte** length, never the character count, and
//! std's `Keccak256` is original Keccak (0x01 padding), not NIST SHA3-256
//! (0x06) — both facts are pinned by vectors.

const std = @import("std");
const secp = @import("secp.zig");

const Keccak256 = std.crypto.hash.sha3.Keccak256;

pub const prefix = "\x19Ethereum Signed Message:\n";

/// The EIP-191 digest of `message`.
pub fn digest(message: []const u8) [32]u8 {
    var st = Keccak256.init(.{});
    st.update(prefix);
    var len_buf: [20]u8 = undefined;
    const len_str = std.fmt.bufPrint(&len_buf, "{d}", .{message.len}) catch unreachable;
    st.update(len_str);
    st.update(message);
    var out: [32]u8 = undefined;
    st.final(&out);
    return out;
}

/// Sign `message` verbatim with `personal_sign` semantics: RFC 6979
/// deterministic nonce, low-S, v ∈ {27, 28}.
pub fn personalSign(message: []const u8, key: secp.PrivateKey) secp.Signature {
    return secp.signDigest(digest(message), key);
}

/// Sign and return the wire form: `"0x"` + 130 lowercase hex chars.
pub fn personalSignHex(message: []const u8, key: secp.PrivateKey) [132]u8 {
    return personalSign(message, key).toWireHex();
}

/// Verify by recovery: the recovered address must equal `expected_address`
/// (20 raw bytes; comparison is byte-wise, i.e. lowercase semantics).
pub fn verify(message: []const u8, sig: secp.Signature, expected_address: [20]u8) bool {
    const recovered = secp.recoverAddress(digest(message), sig) catch return false;
    return std.mem.eql(u8, &recovered, &expected_address);
}

// ---------------------------------------------------------------------------

const testing = std.testing;

test "digest prefixes the UTF-8 byte length, not the character count" {
    // "🍓 strawberry" is 12 code points but 15 UTF-8 bytes; the vector digest
    // for it is checked in tests/vectors.zig. Here: two messages of equal
    // character count but different byte count must not collide the way a
    // char-count implementation would make them.
    const ascii = "aa"; // 2 chars, 2 bytes
    const uni = "é"; // 1 char, 2 bytes
    try testing.expect(!std.mem.eql(u8, &digest(ascii), &digest(uni)));

    // Manual reconstruction for a unicode message.
    const msg = "🍓 strawberry";
    try testing.expectEqual(@as(usize, 15), msg.len);
    var manual = Keccak256.init(.{});
    manual.update("\x19Ethereum Signed Message:\n15");
    manual.update(msg);
    var expected: [32]u8 = undefined;
    manual.final(&expected);
    try testing.expectEqualSlices(u8, &expected, &digest(msg));
}

test "sign/verify round trip and tamper rejection" {
    const key = try secp.PrivateKey.fromHex("0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80");
    const msg = "hello world";
    const sig = personalSign(msg, key);
    try testing.expect(verify(msg, sig, key.address()));
    try testing.expect(!verify("hello world!", sig, key.address()));
    const other = try secp.PrivateKey.fromHex("0x" ++ "33" ** 32);
    try testing.expect(!verify(msg, sig, other.address()));
}

File: app/cli/typescript/src/eip191.ts
/**
 * EIP-191 `personal_sign`: digest, deterministic signing (RFC 6979, low-S,
 * `v = recid + 27`) and address recovery.
 *
 * Byte-exact against `protocol-v1.json` → `eip191[]`.
 */

import * as secp from "@noble/secp256k1";
import { hmac } from "@noble/hashes/hmac";
import { sha256 } from "@noble/hashes/sha2";
import { keccak_256 } from "@noble/hashes/sha3";
import { addressFromPublicKey } from "./wallet.js";
import { bytesToHex, concatBytes, hexToBytes, utf8Bytes } from "./hex.js";

// noble-secp256k1 v2 needs a synchronous HMAC-SHA256 wired in for RFC 6979.
secp.etc.hmacSha256Sync = (key, ...msgs) =>
  hmac(sha256, key, secp.etc.concatBytes(...msgs));

const HALF_N = secp.CURVE.n >> 1n;

/**
 * keccak256( 0x19 || "Ethereum Signed Message:\n" || decimal(utf8_byte_len) || utf8(msg) ).
 *
 * The length is the UTF-8 **byte** length, not the character count.
 */
export function eip191Digest(message: string): Uint8Array {
  const body = utf8Bytes(message);
  const prefix = utf8Bytes(`\x19Ethereum Signed Message:\n${body.length}`);
  return keccak_256(concatBytes(prefix, body));
}

/**
 * Sign a message with EIP-191 personal_sign. Deterministic (RFC 6979, no
 * extra entropy), low-S normalized, `v = recovery_id + 27`.
 *
 * @returns `0x` + 130 lowercase hex (`r(32) || s(32) || v(1)`).
 */
export function personalSign(message: string, privateKey: Uint8Array): string {
  const digest = eip191Digest(message);
  const sig = secp.sign(digest, privateKey, {
    lowS: true,
    extraEntropy: false,
  });
  if (sig.recovery !== 0 && sig.recovery !== 1) {
    throw new Error(`unexpected recovery id ${sig.recovery}`);
  }
  const v = sig.recovery + 27;
  return `0x${bytesToHex(sig.toCompactRawBytes())}${v.toString(16).padStart(2, "0")}`;
}

export interface ParsedSignature {
  r: bigint;
  s: bigint;
  /** 0 or 1. */
  recovery: number;
}

/**
 * Parse a wire signature: `0x` + 130 hex. Emits 27/28 for `v` but tolerates
 * 0/1 on input. Rejects malformed hex, out-of-range `r`/`s` and — because
 * they are malleable — high-S signatures.
 */
export function parseSignature(signature: string): ParsedSignature {
  if (!/^0x[0-9a-fA-F]{130}$/.test(signature)) {
    throw new Error("signature must be 0x + 130 hex characters");
  }
  const bytes = hexToBytes(signature.slice(2));
  const r = secp.etc.bytesToNumberBE(bytes.subarray(0, 32));
  const s = secp.etc.bytesToNumberBE(bytes.subarray(32, 64));
  const v = bytes[64]!;
  let recovery: number;
  if (v === 27 || v === 28) recovery = v - 27;
  else if (v === 0 || v === 1) recovery = v;
  else
    throw new Error(
      `invalid recovery byte ${v} (expected 27/28, tolerating 0/1)`,
    );
  if (r <= 0n || r >= secp.CURVE.n) throw new Error("signature r out of range");
  if (s <= 0n || s >= secp.CURVE.n) throw new Error("signature s out of range");
  if (s > HALF_N) throw new Error("high-S signature rejected (malleable)");
  return { r, s, recovery };
}

/** Recover the lowercase signer address of an EIP-191 signature. */
export function recoverAddress(message: string, signature: string): string {
  const { r, s, recovery } = parseSignature(signature);
  const digest = eip191Digest(message);
  const publicKey = new secp.Signature(r, s)
    .addRecoveryBit(recovery)
    .recoverPublicKey(digest)
    .toBytes(false);
  return addressFromPublicKey(publicKey);
}

/** Verify that `signature` over `message` was made by `expectedAddress`. */
export function verifyPersonalSign(
  message: string,
  signature: string,
  expectedAddress: string,
): boolean {
  try {
    return recoverAddress(message, signature) === expectedAddress.toLowerCase();
  } catch {
    return false;
  }
}
